### PR TITLE
DYN-5319

### PR DESCRIPTION
**Pull request by LongNguyenP**
### Purpose

This PR implements DYN-5319](https://jira.autodesk.com/browse/DYN-5319):  

- Add to the Dynamo Preference Settings a dictionary to store the folder collections that are used by DynamoPlayer or GD as entry-point.  
- Add API for querying, adding, and remove the data from this dictionary.

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

Dynamo Preference Settings now include information about the user-folders in Dynamo Player and Generative Design

### Reviewers

@mjkkirschner @QilongTang @saintentropy

### CODE REVIEW

File: src/DynamoCore/Configuration/PreferenceSettings.cs [hunk 1]
```diff
@@ -680,6 +680,13 @@ public string IronPythonResolveTargetVersion
         public List<string> ReadNotificationIds { get; set; }
         #endregion
 
+        #region Dynamo Player and Generative Design settings
+
+        // Collections of folders used by Dynamo Player or Generative Design as entry points.
+        public Dictionary<string, List<DynamoPlayerFolderItem>> DynamoPlayerFolderItems;
```
Comment (mjkkirschner) [inline src/DynamoCore/Configuration/PreferenceSettings.cs:686]: please add some tests that deserialize / serialize these - there are examples if you search for preference setting tests.

File: src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs [hunk 2]
```diff
@@ -45,7 +45,7 @@
 // to distinguish one build from another. AssemblyFileVersion is specified
 // in AssemblyVersionInfo.cs so that it can be easily incremented by the
 // automated build process.
-[assembly: AssemblyVersion("2.17.0.2866")]
+[assembly: AssemblyVersion("2.17.0.2992")]
```
Comment (mjkkirschner) [inline src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs:48]: please don't commit these files unless there is a reason.

File: src/DynamoCore/Configuration/PreferenceSettings.cs [hunk 3]
```diff
@@ -1097,4 +1104,45 @@ public bool IsCreatedFromValidFile
             get { return isCreatedFromValidFile; }
         }
     }
+
+    public class DynamoPlayerFolderItem
```
Comment (mjkkirschner) [inline src/DynamoCore/Configuration/PreferenceSettings.cs:1108]: thanks for the comments on the members, please add a ///summary for the class as well.

File: src/DynamoCore/Configuration/DynamoPlayerFolder.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+namespace Dynamo.Configuration
+{
+    /// <summary>
+    /// This class describes a folder (usually containing Dynamo graphs) added to the Dynamo Player or Generative Design
+    /// </summary>
+    public class DynamoPlayerFolder
+    {
+        /// <summary>
+        /// The full path if the folder
```
Comment (QilongTang) [inline src/DynamoCore/Configuration/DynamoPlayerFolder.cs:11]: typo `of`

**Comment by QilongTang** [comment]
@LongNguyenP Changes looks fine, but would only guarantee the preference settings serialization/deserialization works, you may need to apply additional changes to make sure setting import/export works, e.g. https://github.com/DynamoDS/Dynamo/pull/13476

File: src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs [hunk 5]
```diff
@@ -45,7 +45,7 @@
 // to distinguish one build from another. AssemblyFileVersion is specified
 // in AssemblyVersionInfo.cs so that it can be easily incremented by the
 // automated build process.
-[assembly: AssemblyVersion("2.17.0.2986")]
+[assembly: AssemblyVersion("2.17.0.3065")]
```
Comment (mjkkirschner) [inline src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs:48]: please avoid commit this, it causes conflicts etc as the file is autogenerated.

File: src/DynamoCore/Configuration/DynamoPlayerFolder.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            DisplayName = displayName;
+            Id = id;
+            IsRemovable = isRemovable;
+            Order = order;
+        }
+
+        /// <summary>
+        /// Create an empty DynamoPlayerFolderItem
+        /// </summary>
+        public DynamoPlayerFolder()
```
[the remark this answers is not in the document]
Comment (saintentropy) [reply src/DynamoCore/Configuration/DynamoPlayerFolder.cs:55]: XML Serializer required the parameterless constructor if you also had another constructor.  Maybe we don't need any constructors.  What do you think @LongNguyenP

[the remark this answers is not in the document]
Comment (LongNguyenP) [reply src/DynamoCore/Configuration/DynamoPlayerFolder.cs]: We have removed all constructors after realising they are not necessary

File: src/DynamoCore/Configuration/PreferenceSettings.cs (hunk 1 above)
Comment (LongNguyenP) [reply src/DynamoCore/Configuration/PreferenceSettings.cs]: Done :)

**Pull request merged**